### PR TITLE
Fix "Unexpected number of outputs after override_all_outputs"

### PR DESCRIPTION
**Pull request by dkozykowski**
### Details:
 - Fix "Unexpected number of outputs after override_all_outputs"

### Tickets:
 - 72669

### CODE REVIEW

File: src/frontends/onnx/frontend/src/input_model.cpp [hunk 1]
```diff
@@ -148,12 +149,69 @@ std::shared_ptr<Function> InputModelONNX::convert() {
...
+        }
+    } else if (const auto output_edge = std::dynamic_pointer_cast<PlaceOutputEdgeONNX>(place)) {
+        if (auto tensor = std::dynamic_pointer_cast<PlaceInputEdgeONNX>(output_edge->get_target_tensor())) {
+            return m_editor->is_correct_tensor_name(tensor->get_names()[0]);
+        }
+    }
+    return false;
+}
+
+std::string InputModelONNX::get_place_name(const ov::frontend::Place::Ptr& place) const {
```
Comment (mbencer) [inline src/frontends/onnx/frontend/src/input_model.cpp:169]: It is not correct. Each kind of `Place` like `PlaceInputEdge` has its own `get_names` implementation. A name of ONNX tensor is only applicable for `PlaceTensor`

File: src/frontends/onnx/frontend/src/input_model.cpp [hunk 2]
```diff
@@ -148,12 +149,69 @@ std::shared_ptr<Function> InputModelONNX::convert() {
...
+        }
+    } else if (const auto output_edge = std::dynamic_pointer_cast<PlaceOutputEdgeONNX>(place)) {
+        if (auto tensor = std::dynamic_pointer_cast<PlaceInputEdgeONNX>(output_edge->get_target_tensor())) {
+            return tensor->get_names()[0];
+        }
+    }
+    throw std::invalid_argument("Incorrect place type");
+}
+
 void InputModelONNX::override_all_outputs(const std::vector<Place::Ptr>& outputs) {
```
Comment (mbencer) [inline src/frontends/onnx/frontend/src/input_model.cpp:186]: Let's try to do sth like:
```
void InputModel::override_all_outputs(const std::vector<Place::Ptr>& outputs) {

    std::vector<Place::Ptr> expected_valid_outputs;
    for (const auto& output : expected_valid_outputs) {
        bool is_correct = is_coorect_place(output);
        if (!is_correct)
            NGRAPH_WARN << "Name  " << output->get_names().at(0),
                        << " of output node is not a correct node name. Ignoring this parameter.";
        else
            expected_valid_outputs.push_back(output);
    }

    extract_subgraph({}, expected_valid_outputs);

    NGRAPH_CHECK(std::all_of(std::begin(valid_outputs),
                             std::end(valid_outputs),
                             [](const Place::Ptr& place) {
                                 return place->is_output();
                             }),
                 "Not all provided arguments of override_all_outputs are new outputs of the model");

    const auto current_outputs = get_outputs();
    NGRAPH_CHECK(std::all_of(std::begin(current_outputs),
                             std::end(current_outputs),
                            [](const Place::Ptr& current_out) {
                                return std::find_if(std::begin(expected_valid_outputs), std::end(expected_valid_outputs)
                                [&current_out](const Place::Ptr& expected_out) {
                                    expected_out->is_equal(current_out);
                                }) != std::end(current_outputs);
                            }),
                "Some other than expected outputs were created during override_all_outputs");
}
```

Comment (mbencer) [inline src/frontends/onnx/frontend/src/input_model.cpp:186]: Try to apply the equivalent solution to `override_all_inputs`

File: src/frontends/onnx/frontend/src/input_model.cpp (hunk 1 above)
Comment (dkozykowski) [reply src/frontends/onnx/frontend/src/input_model.cpp]: Removed this part of the code.

File: src/frontends/onnx/frontend/src/input_model.cpp (hunk 2 above)
Comment (dkozykowski) [reply src/frontends/onnx/frontend/src/input_model.cpp:169]: Updated the code to use your suggestion.

Comment (dkozykowski) [reply src/frontends/onnx/frontend/src/input_model.cpp:169]: Done.

File: src/bindings/python/tests/test_frontend/test_frontend_onnx_editor.py [hunk 3]
```diff
@@ -91,6 +101,23 @@ def create_test_onnx_models():
...
+    add_2 = onnx.helper.make_node("Add", inputs=["in1", "in2"], outputs=["out1"], name="onnx_add_op")
+    relu_2 = onnx.helper.make_node("Relu", inputs=["in2"], outputs=["out2"])
+
+    input_tensors = [
+        make_tensor_value_info("in1", onnx.TensorProto.FLOAT, (2, 2)),
+        make_tensor_value_info("in2", onnx.TensorProto.FLOAT, (2, 2)),
+    ]
+    output_tensors = [
+        make_tensor_value_info("out1", onnx.TensorProto.FLOAT, (1, 2)),
+        make_tensor_value_info("out1", onnx.TensorProto.FLOAT, (2, 2)),
```
Comment (tomdol) [inline src/bindings/python/tests/test_frontend/test_frontend_onnx_editor.py:114]: Can you run ONNX checker on this model? `out` is defined twice but each time with a different shape. I'm not sure that's allowed by the standard.

File: src/bindings/python/tests/test_frontend/test_frontend_onnx_editor.py [hunk 4]
```diff
@@ -346,6 +386,7 @@ def compare_functions(current, expected):  # noqa: C901 the function is too comp
         result = False
         msg += "Not equal number of ops. "
         msg += f"Current: {len(current_ops)}, expected: {len(expected_ops)}. "
+        return result
```
Comment (tomdol) [inline src/bindings/python/tests/test_frontend/test_frontend_onnx_editor.py:389]: If you just return early here, the error message won't be printed.

File: src/frontends/onnx/frontend/src/input_model.cpp [hunk 5]
```diff
@@ -206,28 +207,90 @@ std::shared_ptr<Model> InputModel::convert() {
 }
 
 // Editor features
+bool InputModel::is_coorect_place(const ov::frontend::Place::Ptr& place) const {
```
Comment (tomdol) [inline src/frontends/onnx/frontend/src/input_model.cpp:210]:
```suggestion
bool InputModel::is_correct_place(const ov::frontend::Place::Ptr& place) const {
```

File: src/frontends/onnx/frontend/src/input_model.cpp [hunk 6]
```diff
@@ -206,28 +207,90 @@ std::shared_ptr<Model> InputModel::convert() {
...
+
 void InputModel::override_all_outputs(const std::vector<ov::frontend::Place::Ptr>& outputs) {
-    extract_subgraph({}, outputs);
-    NGRAPH_CHECK(m_editor->model_outputs().size() == outputs.size(),
-                 "Unexpected number of outputs after override_all_outputs");
-    NGRAPH_CHECK(std::all_of(std::begin(outputs),
-                             std::end(outputs),
+    std::vector<Place::Ptr> expected_valid_outputs;
+    for (const auto& output : outputs) {
+        bool is_correct = is_coorect_place(output);
```
Comment (tomdol) [inline src/frontends/onnx/frontend/src/input_model.cpp:233]:
```suggestion
        bool is_correct = is_correct_place(output);
```

File: src/frontends/onnx/frontend/src/input_model.cpp [hunk 7]
```diff
@@ -206,28 +207,90 @@ std::shared_ptr<Model> InputModel::convert() {
...
+                                                         return expected_out->is_equal(current_out);
+                                                     }) != std::end(current_outputs);
+                             }),
+                 "Some other than expected outputs were created during override_all_outputs");
 }
 
 void InputModel::override_all_inputs(const std::vector<ov::frontend::Place::Ptr>& inputs) {
+    std::vector<Place::Ptr> expected_valid_inputs;
+    for (const auto& input : inputs) {
+        bool is_correct = is_coorect_place(input);
```
Comment (tomdol) [inline src/frontends/onnx/frontend/src/input_model.cpp:266]:
```suggestion
        bool is_correct = is_correct_place(input);
```

File: src/frontends/onnx/frontend/src/input_model.hpp [hunk 8]
```diff
@@ -80,6 +80,7 @@ class InputModel : public ov::frontend::InputModel {
 
 private:
     std::shared_ptr<ov::onnx_editor::ONNXModelEditor> m_editor;
+    bool is_coorect_place(const ov::frontend::Place::Ptr& place) const;
```
Comment (tomdol) [inline src/frontends/onnx/frontend/src/input_model.hpp:83]:
```suggestion
    bool is_correct_place(const ov::frontend::Place::Ptr& place) const;
```

**Review by tomdol — CHANGES_REQUESTED** [review]

File: src/bindings/python/tests/test_frontend/test_frontend_onnx_editor.py (hunk 3 above)
Comment (dkozykowski) [reply src/bindings/python/tests/test_frontend/test_frontend_onnx_editor.py:114]: It was a typo, thank you for catching it. Set both shapes to (2, 2).

File: src/bindings/python/tests/test_frontend/test_frontend_onnx_editor.py (hunk 4 above)
Comment (dkozykowski) [reply src/bindings/python/tests/test_frontend/test_frontend_onnx_editor.py]: Removed redundant return.

**Review by tomdol — APPROVED** [review]

**Review by mbencer — APPROVED** [review]

**Pull request merged**